File: Nim/lexer/token.nim
import std/strformat
import std/strutils


type TokenType* {.pure.} = enum
  # Literals
  True, False
  Integer, Float
  String, Char
  Identifier
  # Operators
  Equal
  # Others
  EndOfFile
  Symbol # A generic symbol
  NoMatch # Used internally by the symbol table

type Token* = ref object
  kind*: TokenType              # Type of the token
  lexeme*: string               # The lexeme associated to the token
  line*: int                    # The line where the token appears
  pos*: tuple[start, stop: int] # The absolute position in the source file
                                # (0-indexed and inclusive at the beginning)
  spaces*: int                  # Number of spaces before this token

proc endOfFileToken*(line: int, current: int): Token = 
  Token(kind: EndOfFile, lexeme: "", line: line, pos: (current, current))


proc `$`*(self: Token): string =
  ## Strinfifies
  if self != nil:
    result = &"lexeme={self.lexeme.escape()}, Token(kind={self.kind}, line={self.line}, pos=({self.pos.start}, {self.pos.stop}), spaces={self.spaces})"
  else:
    result = "nil"

proc `$`*(tokens: seq[Token]): string =
  ## Strinfifies
  for tok in tokens:
    result.add $tok
    result.add "\n"

proc `==`*(a, b: Token): bool =
  a.kind == b.kind and a.lexeme == b.lexeme

File: NivaK/Niva/src/main/kotlin/frontend/meta/Token.kt
package frontend.meta

enum class TokenType {
    True, False,

    // Literal types
    Integer, Float, String, Identifier, NullableIdentifier,
    Binary, Octal, Hex, Char,

    // Keywords
    Alias, Type, Union, Use, Constructor, Mut,
    Return, // ^
    ReturnArrow, // -> for return types
    AssignArrow, // <-


    Pipe, // |
    PipeOperator, // |>


    Then, // =>
    Switch, // |
    Else, // |=>

    // brackets
    OpenParen, CloseParen, // ()
    OpenBrace, CloseBrace, // {}
    OpenBracket, CloseBracket, // []

    OpenBraceHash, // #{
    OpenParenHash, // #(

    // punctuation
    Dot, Cascade, Comma, Colon, Apostrophe,// . ; , `


    EndOfFile,
    NoMatch,   // Used internally by the symbol table
    Comment,   // Useful for documentation comments, pragmas, etc.
    BinarySymbol,    // A generic symbol
    Pragma,
    Assign, Equal, NotEqual, // =, ==, !=
    DoubleColon,
    EndOfLine // \n and next line starts not from the dot
    ,
}

data class Position(val start: Int, val end: Int)

data class Token(
    val kind: TokenType,
    val lexeme: String,
    val line: Int,
    val pos: Position,
    val relPos: Position,
    val spaces: Int = 0
) {

    override fun equals(other: Any?): Boolean =
        other is Token && kind == other.kind

    override fun hashCode(): Int {
        var result = kind.hashCode()
        result = 31 * result + lexeme.hashCode()
        result = 31 * result + line
        result = 31 * result + pos.hashCode()
        result = 31 * result + relPos.hashCode()
        result = 31 * result + spaces
        return result
    }
}

fun Token.isIdentifier() = this.kind == TokenType.Identifier || this.kind == TokenType.NullableIdentifier
fun Token.isNullable() = this.kind == TokenType.NullableIdentifier
